### DISCUSSION TITLE
[v3] Custom hook & queryClient

### DISCUSSION THREAD

**Discussion by jzecca** [TanStack/query#1408] [Q&A] [answered]
Hi,

In the case of a custom hook that needs to update the cache, what is the recommended way of accessing the `queryClient`?

- Should I `import { queryClient } from '../wherever/my/client/lives'`
- Or is it safe to use the new `useQueryClient` hook?

Thanks in advance, and thanks to all for the great work :+1:

**Comment by TkDodo** [comment, marked as answer]
since you have to have a `QueryCientProvider` at the top of your react tree with v3, `useQueryClient` will give you that cache. The direct import is only recommended for when you can't use the hook. So unless you need that, I wouldn't even export the client and always retrieve it via the hook. This will make testing easier because you can just wrap your tests with a different Provider and it will just work™️

**Reply by jzecca** [reply]
Thanks a lot for the answer!
